sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala

```diff
@@ -143,4 +144,25 @@ object InternalRow {
     case u: UserDefinedType[_] => getAccessor(u.sqlType)
     case _ => (input, ordinal) => input.get(ordinal, dataType)
   }
+
+  /**
+   * Returns a writer for an `InternalRow` with given data type.
+   */
+  def getWriter(ordinal: Int, dt: DataType): (InternalRow, Any) => Unit = dt match {
+    case BooleanType => (input, v) => input.setBoolean(ordinal, v.asInstanceOf[Boolean])
+    case ByteType => (input, v) => input.setByte(ordinal, v.asInstanceOf[Byte])
+    case ShortType => (input, v) => input.setShort(ordinal, v.asInstanceOf[Short])
+    case IntegerType | DateType => (input, v) => input.setInt(ordinal, v.asInstanceOf[Int])
+    case LongType | TimestampType => (input, v) => input.setLong(ordinal, v.asInstanceOf[Long])
+    case FloatType => (input, v) => input.setFloat(ordinal, v.asInstanceOf[Float])
+    case DoubleType => (input, v) => input.setDouble(ordinal, v.asInstanceOf[Double])
+    case DecimalType.Fixed(precision, _) =>
+      (input, v) => input.setDecimal(ordinal, v.asInstanceOf[Decimal], precision)
+    case CalendarIntervalType | BinaryType | _: ArrayType | StringType | _: StructType |
+         _: MapType | _: ObjectType =>
+      (input, v) => input.update(ordinal, v)
+    case udt: UserDefinedType[_] => getWriter(ordinal, udt.sqlType)
+    case NullType => (input, _) => input.setNullAt(ordinal)
+    case _ => throw new SparkException(s"Unsupported data type $dt")
```

Comment:
one minor point: the codegen version just call `row.update` for un-caught types, which means it supports object type as well. Shall we follow?

Reply:
ok